ui-v2/app/mixins/with-searching.js

```diff
@@ -0,0 +1,32 @@
+import Mixin from '@ember/object/mixin';
+import { inject as service } from '@ember/service';
+import { get, set } from '@ember/object';
+import WithListeners from 'consul-ui/mixins/with-listeners';
+/**
+ * WithSearching mostly depends on a `searchParams` object which must be set
+ * inside the `init` function. The naming and usage of this is modelled on
+ * `queryParams` but in contrast cannot _yet_ be 'hung' of the Controller
+ * object, it MUST be set in the `init` method.
+ * Reasons: As well as producing a eslint error, it can also be 'shared' amongst
+ * child Classes of the component. It is not clear _yet_ whether mixing this in
+ * avoids this and is something to be looked at in future to slightly improve DX
+ * Please also see:
+ * https://emberjs.com/api/ember/2.12/classes/Ember.Object/properties?anchor=mergedProperties
+ *
+ */
+export default Mixin.create(WithListeners, {
+  builder: service('search'),
+  init: function() {
+    this._super(...arguments);
+    const params = this.searchParams || {};
+    this.searchables = {};
+    Object.keys(params).forEach(type => {
+      const key = params[type];
+      this.searchables[type] = get(this, 'builder').searchable(type);
```

Comment:
Had a bit more thought on this, a wise man once said sometimes explaining things by writing some code is better and I'm not the best at explaining things sometimes so this might help explain what my thinking is:

(caveat: complete pseudo code)

```javascript
class ServicesController {
  init() {
    this.search(queryParams)
  }
  search(queryParams) {
    this.items = this._items.filter(
      function() {
        return ...
      }
    )
  }
}
this.route(
  '/ui/services',
  function() {
    return new ServicesController();
  }
);
```

> Why not define the searchFilter on controllers and then wire the filter up in this mixin? Then there is no need for the builder/search service or the initializer

The above does this, the searching is done in the controller. It's really simple, but not at all flexible. The Controller is doing too much, the Controller should just control the interaction between other things, it should just orchestrate things.

Usually I try to follow 4 steps when building things:

1 Separate the concerns

```javascript
class ServicesSearch { // implements Searchable
  search(params) {
    this.items = this._items.filter(
      function() {
        return ...
      }
    )
  }
  add(){...}
}
class Controller {
  init() {
    this.searchable = new ServicesSearch();
    this.searchable.add(this._items);
    this.search(queryParams);
  }
  search(queryParams) {
    this.items = this.searchable.search(queryParams)
  }
}
this.route(
  '/ui/services',
  function() {
    return new ServicesController();
  }
);
```

In my eyes this is better. A searchable is responsible for searching, the Controller just 'uses' the searchable to do the searching. Yes, its slightly more code, but it's much more flexible. I can dynamically use different types of searching and switch between them depending on something (say if I want to search between backend searching and frontend searching for example - something I might end up doing), I can have multiple searches in one Controller if I want. For me though, it's still not ideal.

2 Invert control/configure from the outside

```javascript
class ServicesSearch { // implements Searchable
  search(params) {
    return this.items.filter(
      function() {
        return ...
      }
    )
  }
  add(){...}
}
class Controller {
  constructor(searchable) {
    this.searchable = searchable;
    this.searchable.add(this._items);
    this.search(queryParams);
  }
  search(queryParams) {
    this.items = this.searchable.search(queryParams)
  }
}
this.route(
  '/ui/services',
  function() {
    return new Controller(new ServicesSearch());
  }
);
```

This is a tiny change, but is massively important to me. Notice I only now need one Controller, I can just implement multiple types of searchables. It's extremely flexible, easy to unit test and reason about. I can easily switch different types of searches in and out so I can measure how quick/appropriate different types are during development. I could make a searchable that lets the user dynamically choose between say a RegEx search, or a filtered search, or a backend search. Step 1 and 2 really come into their own when you are testing things, and testing helps with steps 3 and 4.

3 Keep it simple

The `ServicesSearch.search` body is really simple, it's very nice having this live on its own in it's own file. Generally the actual logic of these things is either really simple, or there's something already out there on npm/`<insert package registry here>` that does the logic for you. I'm not that good at the actual logical stuff, looking at what you've been doing with your graphs you're probably far better than me at it!

I do think you are right here though, and where I think I am failing with this step right now in my app is, in my actual app code here (not this pseudo code) it is a little bit complicated, but that is mainly down to the framework and the absence of simple constructor injection. Ember uses property injection and a service locator and everything has to be wrapped in a `Service`, these facts mean you do have to engineer things a little more than usual. Using property injection means public properties, which are likely to be accidentally overwritten at some point, but I've accepted that. An alternative to what I've done would have been to:

```javascript
Controller.extend(
  {
    searchable: service('servicesSearch')
  }
);
Controller.extend(
  {
    searchable: service('nodesSearch')
  }
);
Controller.extend(
  {
    searchable: service('kvSearch')
  }
);
```

But this means wrapping every single filter function in `Service` for no other reason but to be able to inject it. I'd rather not have to do that. So I've essentially used a search specific service builder/container.

```javascript
Controller.extend(
  {
    builder: service('search'),
    init: function() {
      this.searchable = get(this, 'builder').searchable('servicesSearch')
      // more or less, I'm actually doing this with a computed property
      this.searchable.add(this.items);
      this.searchable.search(this.queryParams);
    }
  }
);
// which is an ember-y (but without loads of `Service`s) way of doing this
class Controller {
  constructor(searchable) {
    this.searchable = searchable;
    this.searchable.add(this._items);
    this.search(queryParams);
  }
  search(queryParams) {
    this.items = this.searchable.search(queryParams)
  }
}
```

I have some thoughts on making this simpler, and I'm trying my best to do it so it feels as ember-like as possible, but keep things separate, configurable from the outside, and simple. One caveat to this is, as you do this more and more, all of the setup and injection gets tedious and complicated, which is where a ServiceContainer builder comes in (preferably declarative), ember doesn't have one of those and I've never seen an addon for one.

4 Keep things DRY

This is usually my final step, but generally if you've done the above this just happens naturally. For example here, if I need a completely new type of services page, that requires its own `Controller` in ember. I can just reuse the same `Searchable` class. If I'd left the searching functionality in the `Controller` I wouldn't have been able to do this. You are quite right in saying that right now I don't need to do this, but there is the possibility to reuse in the 'Services' and the 'Node detail > Service Tab' I've just not got there yet.

Hopefully you'll see these 4 steps/ideas in most things I do. For declarative things, slots in HTML/HTMLBars enable me to configure from the outside, and keep things separate. `@extend` in CSS keeps things DRY and lets you configure your selectors from the outside, CSS naturally lets you configure properties from the outside. Anything imperative like javascript or whatever else will use some sort of approach similar to the pseudocode above, just with a different syntax depending on the language.

Apologies if the above comes across a little too much, and I'm by no means saying this is perfect or *the* way to build things, but its my way. Everyone has different ways of approaching problems, the above is generally mine and it might help to explain a lot of my decisions a bit better.